### PR TITLE
add upgrade test for standalone neg

### PR DESCRIPTION
**Pull request by freehan**

### CODE REVIEW

**Comment by freehan** [comment]
cc: @skmatti

**Comment by rramkumar1** [comment]
@freehan I would suggest splitting up this change into two commits. One for the upgrade test interface + modifying the existing test and a second for the standalone neg test.

**Comment by freehan** [comment]
make sense

**Comment by freehan** [comment]
Commits splitted

File: cmd/e2e-test/upgrade/interface.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+import (
+	"k8s.io/ingress-gce/pkg/e2e"
+	"testing"
+)
+
+// UpgradeTest is an interface for writing generic upgrade test.
+// Each test case should implement the following functions and the GenericUpgrade test is able to execute it.
+// TODO: add version compatibility into the interface
+type UpgradeTest interface {
```
Comment (rramkumar1) [inline cmd/e2e-test/upgrade/interface.go]: I would declare this interface in pkg/e2e/interface.go instead. Or reuse an existing file. 

I don't think it makes sense to have library code in pkg/cmd.

File: cmd/e2e-test/upgrade/interface.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+	// Init initialized the Upgrade Test
```
Comment (rramkumar1) [inline cmd/e2e-test/upgrade/interface.go]: Add a period after all these comments.

File: cmd/e2e-test/upgrade_test.go [hunk 3]
```diff
@@ -113,3 +114,51 @@ func TestUpgrade(t *testing.T) {
 		})
 	}
 }
+
+func TestGenericUpgrade(t *testing.T) {
```
Comment (rramkumar1) [inline cmd/e2e-test/upgrade_test.go]: I think you need to add a TODO to migrate the existing upgrade test to this one?

File: cmd/e2e-test/upgrade/interface.go (hunk 1 above)
Comment (freehan) [reply cmd/e2e-test/upgrade/interface.go]: Yeah. I thought about it. I was trying to find a proper home for the lib. But then, I noticed that both the implementation and the caller will be in `cmd/e2e-test`. I might as well put it there. 

Okay. I will put it in pkg/e2e/interface.go for consistency

File: cmd/e2e-test/upgrade_test.go (hunk 3 above)
Comment (freehan) [reply cmd/e2e-test/upgrade_test.go]: done

File: pkg/e2e/interface.go [hunk 4]
```diff
@@ -0,0 +1,35 @@
+/*
```
Comment (rramkumar1) [inline pkg/e2e/interface.go]: I would call this file interfaces.go

File: pkg/e2e/interface.go [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+*/
+
+package e2e
+
+import (
+	"testing"
+)
+
+// UpgradeTest is an interface for writing generic upgrade test.
+// Each test case should implement the following functions and the GenericUpgrade test is able to execute it.
```
Comment (rramkumar1) [inline pkg/e2e/interface.go]: I would just omit this comment. Conceptually, this interface has no knowledge of GenericUpgrade

File: cmd/e2e-test/upgrade/neg.go [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+		Ingress: false,
+		ExposedPorts: map[int32]annotations.NegAttributes{
+			int32(443): {},
+			int32(80):  {},
+		},
+	}
+	expectServicePort = []string{"80", "443"}
+)
+
+type StandaloneNeg struct {
```
Comment (rramkumar1) [inline cmd/e2e-test/upgrade/neg.go]: Add a comment here saying that StandaloneNeg implements e2e.UpgradeTest

File: cmd/e2e-test/upgrade/neg.go [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+	t         *testing.T
+	s         *e2e.Sandbox
+	framework *e2e.Framework
+}
+
+func NewStandaloneNeg() e2e.UpgradeTest {
```
Comment (rramkumar1) [inline cmd/e2e-test/upgrade/neg.go]: NewStandaloneNegUpgradeTest

File: cmd/e2e-test/upgrade/neg.go [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+func (sn *StandaloneNeg) Init(t *testing.T, s *e2e.Sandbox, framework *e2e.Framework) error {
+	sn.t = t
+	sn.s = s
+	sn.framework = framework
+	return nil
+}
+
+func (sn *StandaloneNeg) PreUpgrade() error {
```
Comment (rramkumar1) [inline cmd/e2e-test/upgrade/neg.go]: Add a comment to functions that implement the interface like:

"foo implements bar"

**Comment by freehan** [comment]
Fixed the comments

**Pull request merged**